### DISCUSSION TITLE
Cannot find GENERATING TRAINING DATA

### DISCUSSION THREAD

**Discussion by minjac** [halide/Halide#6060] [Q&A] [answered]
Hi, I have learned Halide for a few weeks, but I do not know how to generate training data randomly for training cost model.
In _Learning to Optimize Halide with Tree Search and Random Programs_, the author says the source code is in the supplemental material, but I cannot found that.
I do not find source code about randomly generating training data in this project. Could you tell me where I could find this source code.
Thanks.

**Comment by abadams** [comment, marked as answer]
The supplemental material for that paper is the branch "standalone_autoscheduler". The random pipeline generator is here: https://github.com/halide/Halide/tree/standalone_autoscheduler/apps/random_pipeline

**Comment by ryanstout** [comment]
@abadams Just curious, any plans to merge in standalone_autoscheduler or standalone_autoscheduler_gpu? I would guess if I can get adams2019 trained on my hardware it would beat any of my hand done schedules. (Also, thanks for all of the hard work to create/maintain Halide!)

@minjac if you do figure out how to train the adams2019, please let me know the details :-)  I've got some armv7 and aarch64 hardware that I've been wanting to train on, but couldn't figure out how to get the training working.

**Reply by minjac** [reply]
I don't know how to train so far, all I know is in the paper _Learning to Optimize Halide with Tree Search and Random Programs_. If I figure that, I will tell you.

**Reply by ryanstout** [reply]
@minjac Did you ever figure out how to train the cost model? Also, funny, I've got armv7 and aarch64 hardware as well I'm running on :-)

**Comment by abadams** [comment]
Adams2019 was merged! It's in src/autoschedulers. If you have specific apps you want to schedule, I'd just use it in autotuning mode rather that worrying about doing a big training job.

**Comment by ryanstout** [comment]
@abadams thanks. I guess I was a bit confused since it looks like the random_pipeline stuff didn't get merged in. Unfortunately, my pipeline is using the python bindings, and I wasn't able to get it work with autotune_loop.sh (since it expects a generator).

**Comment by rioffe** [comment]
@abadams I am revisiting random_pipeline code and scripts with the goal of obtaining a new set of weights for the cascade lakes machines with AVX512 extensions. I am also trying to figure out what would be the best HL_MACHINE_PARAMS triple for these machines should be. For the last one, I am currently picking 8,36608000,40, since these machines have 8 cores, 36608K of L3 cache and the 40 balance parameter is just from the previous settings. It might be worth it to perform a grid search similar to the one done in generate_master_autotuned.sh to find the best setting for the balance parameter.

As for the methodology of obtaining a new set of weights, it seems like there is no script for that. There is a script (autotune_loop.sh) to tune a single random pipeline and get updated weights that way. There are also scripts (e.g. bench_100.sh) to generate and build random pipelines  aw well as benchmark those pipelines and extract results from those runs. But there is no script to create a set of weights by retraining the model weights on all the random pipelines.

What would be the right way to approach this? How was the set of Adams2019 baseline.weights obtained? If I want to recreate this set of weights, how many random pipelines should I generate? How deep should those pipelines be? What should the initial set of weights be? Would random weights be a good start? What should be the retraining procedure?

**Reply by abadams** [reply]
The baseline weights were generated by running lots of instances of the autotune script on the random pipeline generator on a cluster for as many iterations as we could afford, then collecting all the sample files and running the retraining binary on the whole set to get a single set of weights. This set of weights was then redistributed to the machines in the cluster and the whole process repeated about five times. The first set of weights used was just random. I'm not convinced this was a particularly great way to do things. More recent experiments have shown that the weights generated in this way have almost no predictive power on some of the apps. I'd instead run the autotuning loop on all the apps (or some other corpus of code) independently, and collect and combine all those sample files. You want to get several hundred thousand total samples, IIRC.

For the machine params, Adams2019 only pays attention to the core count. The other two factors are ignored.
